### ISSUE TITLE
The way Clion does backspace is to die for

### ISSUE THREAD

**Issue by gilescope** [rust-analyzer/rust-analyzer#4182]
I didn't know that backspace could be improved. How wrong I was! I hope JetBrains roll out Clion backspace handling across all their products. I'd also love to see it in Rust Analyzer.

There's already special enter key handling so backspace should be possible...

I think the rules are that the first backspace takes you up a level in indentation, but then the second gobbles up all the whitespace to the start of the line and the line brake so that it appends onto the end of the previous line. This turns out to be what I wanted to do so much of the time - it feels so natural.

Small things make such a difference. Well done Clion for bringing a better backsapce to the world.

**Comment by matklad** [comment]
Yeah, open line / newline / backspace all can benefit from indentation info... 

There are two pieces to the puzzle:

* implement indentation model, so that we know what the indentation should be. I wonder whether this is actually blocked on formatting? It seems like we can start coding something up *just* for indent, and then re-rail it onto proper formatting engine...
* implement editor-side hooks. We already have a hook for enter, the one for backspace can be added similarly

Heck, I *think* we should have this sooner rater than later, as it seems like our `onEnter` does not stress the editing as much as it should.

**Comment by flodiebold** [comment]
I have to say that the behavior described in the issue seems like something that should be handled by the editor and not be language-specific, but maybe I'm misunderstanding it :thinking:

**Comment by matklad** [comment]
It has to be language specific I think. For example, here's some Kotlin code

```
fun <T> frobnicate(a: T, b: T)
        where T : Comparable<T>,
              T : Runnable {

    frobnicate(a,
               b
    )
}
```

Here, the `T`'s should be aligned according to code style, but `b` should be just indented. The backspace behavior is different for two:

![indent](https://user-images.githubusercontent.com/1711539/80473160-09999e80-8946-11ea-8dd8-53c5b40d2953.gif)

Also, unrelated, but the gutter icons in IntelliJ are soooo much better than code lens :( Like this little recursion swirly it's just cute, and we can't have that in VS Code TT

**Comment by Timmmm** [comment]
I think it would be a bad idea to implement this in VSCode for the same reason that messing with multiline commends was a bad idea:

1. Users won't expect it, and it produces editing inconsistencies with editing different languages.
2. Messing with text editing in an extension causes performance issues in VSCode at the moment.

**Comment by bjorn3** [comment]
> 1. Users won't expect it, and it produces editing inconsistencies with editing different languages.

VSCode already does this. For Rust it often produces unexpected or plain wrong results like the following though:

```rust
let a = b
    .c();<|>
```

Pressing enter gives:

```rust
let a = b
    .c();
    <|>
```

That is too much indentation. Even when you already deindented the next line pressing enter decides that it needs to indent anyway:
```rust
let a = b
    .c();
<|>
```

gives

```rust
let a = b
    .c();

    <|>
```

This is annoying.

**Comment by Timmmm** [comment]
> VSCode already does this.

Not sure exactly what you mean by "this", but it does seem that VSCode does some limited language-specific indentation. It is controlled by `onEnterRules`, e.g. [Python](https://github.com/microsoft/vscode/blob/master/extensions/python/src/pythonMain.ts), [PHP](https://github.com/microsoft/vscode/blob/master/extensions/php-language-features/src/phpMain.ts).

It is regex based, and the actions are limited to the following, plus appending text (e.g. `///` for comments)

```
	export enum IndentAction {
		/**
		 * Insert new line and copy the previous line's indentation.
		 */
		None = 0,
		/**
		 * Insert new line and indent once (relative to the previous line's indentation).
		 */
		Indent = 1,
		/**
		 * Insert two new lines:
		 *  - the first one indented which will hold the cursor
		 *  - the second one at the same indentation level
		 */
		IndentOutdent = 2,
		/**
		 * Insert new line and outdent once (relative to the previous line's indentation).
		 */
		Outdent = 3
	}
```

I don't think it is quite powerful enough to be useful in your example because you can only look back one line.

On reflection I think you are right that it would be nice to fix it, but I don't think the best way is by intercepting ever keypress in the language server because of the performance issues we've already seen.

**Comment by kjeremy** [comment]
As of vscode 1.53.0 `OnEnterRule` now supports: `OnEnterRule.previousLineText` so we can get at the previous line.
